index.js

```diff
@@ -0,0 +1,348 @@
+     * @param  {String}         config.scms[x].config              Configuration to construct the module with
+     * @param  {String}         config.scms[x].config.displayName  Nickname to displaoy of the scm
+     * @return {ScmRouter}
+     */
+    constructor(config = {}) {
+        const ecosystem = config.ecosystem;
+        const scmsConfig = config.scms;
+
+        super();
+
+        this.scms = {};
+
+        if (typeof scmsConfig === 'object') {
+            if (!(Array.isArray(scmsConfig))) {
+                throw new Error('No scm config passed in.');
+            }
+            scmsConfig.forEach((scm) => {
+                if (typeof scm.config === 'object' && typeof scm.config.displayName === 'string') {
+                    const options = hoek.applyToDefaults({ ecosystem }, scm.config);  // Add ecosystem to scm options
+
+                    this.loadPlugin(scm.plugin, options);
+                } else {
+                    throw new Error(`Display name not specified for ${scm.plugin} scm plugin`);
+                }
+            });
+        }
+
+        if (Object.keys(this.scms).length === 0) {
+            throw new Error('No scm config passed in.');
+        }
+    }
+
+    /**
+     * load scm module
+     * @method loadPlugin
+     * @param  {String}         plugin       load plugin name
+     * @param  {Object}         options      settings for scm module
+     */
+    loadPlugin(plugin, options) {
+        if (plugin === 'router') {
+            console.warn('The plugin of scm-router can not be specified for scms setting');
+
+            return;
+        }
+
+        // eslint-disable-next-line global-require, import/no-dynamic-require
+        const ScmPlugin = require(`screwdriver-scm-${plugin}`);
+        const scmPlugin = new ScmPlugin(options);
+        const scmContexts = scmPlugin.getScmContexts();
+        const scmContext = scmContexts[0]; // plugins return only one scmContext
+
+        if (!scmContext || typeof scmContext !== 'string') {
+            console.warn(`Illegal scmContext for ${plugin} scm plugin`);
+
+            return;
+        }
+
+        if (typeof this.scms[scmContext] === 'object') {
+            console.warn(`Duplicate scm config for ${scmContext}`);
+
+            return;
+        }
+
+        this.scms[scmContext] = scmPlugin;
+    }
+
+    /**
+     * choose webhook scm module
+     * @method chooseWebhookScm
+     * @param  {Object}     headers          The request headers associated with the webhook payload
+     * @param  {Object}     payload          The webhook payload received from the SCM service
+     * @return {Promise}                     scm object
+     */
+    chooseWebhookScm(headers, payload) {
+        return new Promise((resolve, reject) => {
+            // choose a webhook scm module, or null if there is no suitable one
+            async.detectSeries(this.scms, (scm, cb) => {
```

Comment:
Are the scms in `this.scms` ordered in any way? If order is not important it might be better to use `detect` instead of `detectSeries` to avoid blocking the event loop.